# Data Mungin

## Step 1: Importing the libraries

In [1]:
import numpy as np
import pandas as pd
import datetime
from datetime import datetime, date

## Step 2: Reading the data

In [2]:
df_fighters = pd.read_csv("/Users/ksah4444/Documents/UFC_ML_Prediction/data/fighter_stats/2022-02-23T11-41-35.csv")


## Step3: Take a glimpse at the data

In [3]:
df_fighters

dob        fighter_id  height  n_draw  n_loss  n_win  \
0     Mar 02, 1992  2af2f2e26c4c0402   5' 8"       0       2      9   
1     Jun 26, 1982  0c277f3ff66b0208   6' 2"       1       6     15   
2     Jun 30, 1976  d29b5c4f22c6357d   6' 2"       1      16     39   
3     Feb 28, 1998  e415ea6f8d41c917  5' 10"       0       3      8   
4     Apr 18, 1996  fdbefee0827e1567   5' 8"       0       4     11   
...            ...               ...     ...     ...     ...    ...   
3736  Aug 08, 1988  29af297d9f1de0f8  5' 11"       1       4     19   
3737  Sep 19, 1982  ad32471f01e7b1a5  5' 10"       0       5      9   
3738  Jun 20, 1978  ffc088e64fab57e9   6' 1"       0      12     37   
3739  Sep 23, 1966  ce783bf73b5131f9   5' 8"       1       9     15   
3740            --  13b2f59210dda9cc  5' 10"       0       2      4   

                 name reach  sig_str_abs_pM  sig_str_def_pct  sig_str_land_pM  \
0        Hunter Azure   69"            2.08             0.58             3.92   
1      Virgil Zwicker   74"            4.87             0.39             3.34   
2        Gilbert Yvel   77"            1.78             0.50             1.05   
3             Xie Bin   72"            2.60             0.63             2.20   
4            Wu Yanan   66"            4.82             0.51             4.51   
...               ...   ...             ...              ...              ...   
3736    Damon Jackson   71"            3.25             0.48             2.54   
3737   Jeremy Jackson    --            2.03             0.25             0.71   
3738  Quinton Jackson   73"            2.51             0.55             3.25   
3739   Eugene Jackson    --            3.44             0.32             2.19   
3740    Kevin Jackson    --            0.00             0.00             0.00   

      sig_str_land_pct    stance  sub_avg  td_avg  td_def_pct  td_land_pct  \
0                 0.53  Orthodox      1.3    1.97        0.67         0.34   
1                 0.48       NaN      0.0    1.31        0.50         0.30   
2                 0.47  Orthodox      1.0    0.00        0.25         0.00   
3                 0.38  Orthodox      0.0    4.00        0.00         0.44   
4                 0.45  Orthodox      0.3    0.61        0.66         0.22   
...                ...       ...      ...     ...         ...          ...   
3736              0.38    Switch      2.5    2.48        0.35         0.33   
3737              0.27  Orthodox      0.0    1.52        0.58         0.40   
3738              0.54  Orthodox      0.1    1.52        0.73         0.56   
3739              0.33  Orthodox      2.3    1.91        0.16         0.71   
3740              0.00    Switch      0.0    0.00        0.00         0.00   

        weight  
0     145 lbs.  
1     205 lbs.  
2     225 lbs.  
3     145 lbs.  
4     135 lbs.  
...        ...  
3736  145 lbs.  
3737  170 lbs.  
3738  205 lbs.  
3739  185 lbs.  
3740  199 lbs.  

[3741 rows x 18 columns]

In [4]:
df_fighters.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3741 entries, 0 to 3740
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   dob               3741 non-null   object 
 1   fighter_id        3741 non-null   object 
 2   height            3741 non-null   object 
 3   n_draw            3741 non-null   int64  
 4   n_loss            3741 non-null   int64  
 5   n_win             3741 non-null   int64  
 6   name              3741 non-null   object 
 7   reach             3741 non-null   object 
 8   sig_str_abs_pM    3741 non-null   float64
 9   sig_str_def_pct   3741 non-null   float64
 10  sig_str_land_pM   3741 non-null   float64
 11  sig_str_land_pct  3741 non-null   float64
 12  stance            2933 non-null   object 
 13  sub_avg           3741 non-null   float64
 14  td_avg            3741 non-null   float64
 15  td_def_pct        3741 non-null   float64
 16  td_land_pct       3741 non-null   float64


## Step 4: Clean df_fighters

| Column  | Description |                                                                                   
| ------------- | ------------- |                               
| 	dob  | Date of birth  |
| fighter_id  | Randmly generated ID number  |
| 	height  | height in ft'inch"  |
| weight  | weight class in lbs  |
| reach  |  length measuring from the tip of the middle finger on one hand to the other|
| n_draw  | Number of draws  |
| 	n_loss  | Number of losses  |
| n_win  | Number of wins  |
| 	name  | Name of the fighter  |
| sig_str_abs_pM  | Significant Strikes Landed per Minute  |
| 	sig_str_def_pct  | Significant Striking Accuracy  |
| sig_str_land_pM  | Significant Strikes Absorbed per Minute  |
| 	sig_str_land_pct  | Significant Strike Defence (the % of opponents strikes that did not land)  |
| sub_avg  | Average Submissions Attempted per 15 minutes  |
| td_avg  | Average Takedowns Landed per 15 minutes |
| td_def_pct  | Takedown Defense (the % of opponents TD attempts that did not land)  |
| td_land_pct  | Takedown Accuracy  |

In [5]:
# First let's check if there is any missing values in the name column
null_name = pd.isnull(df_fighters["name"])
df_fighters[null_name]

Empty DataFrame
Columns: [dob, fighter_id, height, n_draw, n_loss, n_win, name, reach, sig_str_abs_pM, sig_str_def_pct, sig_str_land_pM, sig_str_land_pct, stance, sub_avg, td_avg, td_def_pct, td_land_pct, weight]
Index: []

In [6]:
# Rearrange columns now (for presentation screen shot)
df_fighters = df_fighters[['dob', 'fighter_id','height','reach', 'name', 
       'weight','n_draw','n_loss', 'n_win',
       'sig_str_abs_pM', 'sig_str_def_pct', 
       'sig_str_land_pM','sig_str_land_pct' ,
       'stance', 'sub_avg', 'td_avg',
       'td_def_pct','td_land_pct']]

In [7]:
# Let's check if there is any duplicates
df_fighters[df_fighters.duplicated(subset="name", keep=False)]

dob        fighter_id  height reach              name  \
553   Jan 15, 1991  d0314416a7f26527   5' 9"   70"  Michael McDonald   
555   Feb 06, 1965  d52ef694108f8235  5' 11"    --  Michael McDonald   
1296  Jul 13, 1989  12ebd7d157e91701   6' 0"   74"       Bruno Silva   
1297  Mar 16, 1990  294aa73dbf37d281   5' 4"   65"       Bruno Silva   
2952  Oct 07, 1992  fb3e61720be4690c   6' 0"   72"        Mike Davis   
2957            --  c8661e204c66f325      --    --        Mike Davis   
3298  Aug 29, 1989  3a28e1e641366308  5' 10"   71"        Joey Gomez   
3300  Jul 21, 1986  0778f94eb5d588a5  5' 10"   73"        Joey Gomez   
3686            --  a45bab49951a45cd   6' 1"    --      Tony Johnson   
3694  May 02, 1983  3641a0d117e9bc6c   6' 2"   76"      Tony Johnson   

        weight  n_draw  n_loss  n_win  sig_str_abs_pM  sig_str_def_pct  \
553   135 lbs.       0       4     17            2.76             0.57   
555   205 lbs.       0       1      1            0.40             0.50   
1296  185 lbs.       0       6     22            2.87             0.32   
1297  125 lbs.       2       5     12            3.23             0.58   
2952  155 lbs.       0       2      9            6.76             0.57   
2957        --       0       0      2            0.00             0.00   
3298  155 lbs.       0       1      7            3.33             0.50   
3300  135 lbs.       0       2      6            4.46             0.55   
3686  265 lbs.       0       3     11            4.73             0.31   
3694  205 lbs.       0       2      7            3.67             0.22   

      sig_str_land_pM  sig_str_land_pct    stance  sub_avg  td_avg  \
553              2.69              0.42  Orthodox      1.4    1.09   
555              0.00              0.00  Orthodox      0.0    0.00   
1296             5.74              0.60  Orthodox      0.0    0.00   
1297             2.98              0.46  Orthodox      0.0    2.89   
2952             6.24              0.53  Orthodox      0.0    1.39   
2957             0.00              0.00       NaN      0.0    0.00   
3298             3.73              0.49  Orthodox      0.0    2.00   
3300             2.44              0.28  Orthodox      0.0    0.62   
3686             2.00              0.53       NaN      0.0    2.00   
3694             4.00              0.92  Orthodox      0.0    0.00   

      td_def_pct  td_land_pct  
553         0.52         0.66  
555         0.00         0.00  
1296        0.68         0.00  
1297        0.64         0.31  
2952        0.69         0.33  
2957        0.00         0.00  
3298        0.00         0.28  
3300        0.50         1.00  
3686        0.00         0.22  
3694        0.90         0.00

#### It seems we have 5 identical name fighters:**Michael McDonald**, **Bruno Silva**, **Mike Davis**, **Joey Gomez**, and **Tony Johnson**. 
- **Michael McDonald**: There is a 70lbs (30kg) difference in weight, so we can confidently say they are not the same person
- **Bruno Silva**: Again there is a 60lbs (25kg) difference in weight, not to mention the 8" height differnce
- **Mike Davis**: The second value has no info, so we can confidently drop this row
- **Joey Gomez**: This one is a bit tricky because they same height, and a sligh weight difference (20lbs). But apon further research it turns they are different people 
- **Tony Johnson**: Another trick one because of the very slight height differnce (1"), and closness in weight classes. But apon further research it turns they are different people.

http://www.espn.com/mma/fighters

### Result: 
- We only drop the second **Mike Davis** row, because of it's lack of info
- Add the weight class of other fightes to thier names
#### I didn't use other columns such as fighter_id because I didn't know how reliable they are

In [8]:
# Change the name of duplicated fighters, so there would'nt be confusion in later processes
df_fighters.iloc[555, 4] = "Michael McDonald 205"
df_fighters.iloc[1297, 4] = "Bruno Silva 125"
df_fighters.iloc[3300, 4] = "Joey Gomez 135"
df_fighters.iloc[3686, 4] = "Tony Johnson 265"

In [9]:
# Drop "Mike Davis" Second Value
df_fighters = df_fighters.drop(labels=2957, axis=0).reset_index()

In [10]:
# Now drop the fighter id as we have no need for it.
df_fighters.drop(['fighter_id'], axis = 1, inplace=True)

In [11]:
# Instead use "name" column as an index
df_fighters.set_index("name", inplace=True)

### Before this, I have been going through the data in a column by column process. Now, I want to skip the **record** and **dob** to get a look at the **striking stats**. Because without the **striking stats**, our model won't have enough features to go through.

In [12]:
# Now let's see how many fighters are missing both or either of the striking and grappling stats
df_st = df_fighters.loc[(df_fighters['sig_str_abs_pM'] == 0) & (df_fighters['sig_str_def_pct'] == 0) & (df_fighters['sig_str_land_pM'] == 0)
& (df_fighters['sig_str_land_pct'] == 0)]
df_gr = df_fighters.loc[(df_fighters['sub_avg'] == 0) & (df_fighters['td_avg'] == 0) & (df_fighters['td_def_pct'] == 0) & (df_fighters['td_land_pct'] == 0)]
df_st_gr = df_fighters.loc[(df_fighters['sub_avg'] == 0) & (df_fighters['td_avg'] == 0) & (df_fighters['td_def_pct'] == 0) & (df_fighters['td_land_pct'] == 0)
& (df_fighters['sig_str_abs_pM'] == 0) & (df_fighters['sig_str_def_pct'] == 0) & (df_fighters['sig_str_land_pM'] == 0)
& (df_fighters['sig_str_land_pct'] == 0)]
#print(len(df_st)/len(df_fighters))
#print(len(df_gr)/len(df_fighters))
print(len(df_fighters))
print(len(df_st_gr))
print(len(df_st_gr)/len(df_fighters))
df_st_gr.head(5)


3740
642
0.1716577540106952


index           dob height reach    weight  n_draw  n_loss  \
name                                                                         
Vitaly Minakov     14  Feb 06, 1985  6' 2"    --  249 lbs.       0       0   
Kaleo Kwan         18  Aug 01, 1970  5' 7"    --  155 lbs.       0      16   
John Kuhner        35  Oct 12, 1982     --    --  170 lbs.       0       2   
Jorgen Kruth       39  May 08, 1974  6' 2"    --  205 lbs.       0       0   
Marcus Kowal       46            --     --    --  135 lbs.       1       1   

                n_win  sig_str_abs_pM  sig_str_def_pct  sig_str_land_pM  \
name                                                                      
Vitaly Minakov     20             0.0              0.0              0.0   
Kaleo Kwan         11             0.0              0.0              0.0   
John Kuhner         1             0.0              0.0              0.0   
Jorgen Kruth        5             0.0              0.0              0.0   
Marcus Kowal        3             0.0              0.0              0.0   

                sig_str_land_pct stance  sub_avg  td_avg  td_def_pct  \
name                                                                   
Vitaly Minakov               0.0    NaN      0.0     0.0         0.0   
Kaleo Kwan                   0.0    NaN      0.0     0.0         0.0   
John Kuhner                  0.0    NaN      0.0     0.0         0.0   
Jorgen Kruth                 0.0    NaN      0.0     0.0         0.0   
Marcus Kowal                 0.0    NaN      0.0     0.0         0.0   

                td_land_pct  
name                         
Vitaly Minakov          0.0  
Kaleo Kwan              0.0  
John Kuhner             0.0  
Jorgen Kruth            0.0  
Marcus Kowal            0.0

### I'm really struggling to make sense of these numbers, but I need to move on (it seems very "venn diagramy")

### It seems around %20 percent of our fighters are missing both striking, and grappling stats. So we can drop them, to make sure our model runs throught good features

In [13]:
df_fighters_cl = df_fighters.loc[~((df_fighters['sig_str_abs_pM'] == 0) &
                               (df_fighters['sig_str_def_pct'] == 0) & 
                               (df_fighters['sig_str_land_pM'] == 0) &
                               (df_fighters['sig_str_land_pct'] == 0) &
                               (df_fighters['sub_avg'] == 0) &
                               (df_fighters['td_avg'] == 0) &
                               (df_fighters['td_def_pct'] == 0) &
                               (df_fighters['td_land_pct'] == 0))].copy()
print(len(df_fighters_cl)/len(df_fighters))
print(len(df_fighters) - len(df_fighters_cl))
df_fighters_cl.head(1)

0.8283422459893048
642


index           dob height reach    weight  n_draw  n_loss  \
name                                                                       
Hunter Azure      0  Mar 02, 1992  5' 8"   69"  145 lbs.       0       2   

              n_win  sig_str_abs_pM  sig_str_def_pct  sig_str_land_pM  \
name                                                                    
Hunter Azure      9            2.08             0.58             3.92   

              sig_str_land_pct    stance  sub_avg  td_avg  td_def_pct  \
name                                                                    
Hunter Azure              0.53  Orthodox      1.3    1.97        0.67   

              td_land_pct  
name                       
Hunter Azure         0.34

### Now we can move on the **dob** column, to calculate the age of a fighter

In [14]:
# First, let's seperate the month, day, and year to calculate the age
df_fighters_cl[["Month", "Day", "Year"]] = df_fighters_cl["dob"].str.split(" ", expand=True)
df_fighters_cl

index           dob  height reach    weight  n_draw  n_loss  \
name                                                                           
Hunter Azure         0  Mar 02, 1992   5' 8"   69"  145 lbs.       0       2   
Virgil Zwicker       1  Jun 26, 1982   6' 2"   74"  205 lbs.       1       6   
Gilbert Yvel         2  Jun 30, 1976   6' 2"   77"  225 lbs.       1      16   
Xie Bin              3  Feb 28, 1998  5' 10"   72"  145 lbs.       0       3   
Wu Yanan             4  Apr 18, 1996   5' 8"   66"  135 lbs.       0       4   
...                ...           ...     ...   ...       ...     ...     ...   
Mike Jackson      3734  Mar 22, 1985   6' 2"    --  170 lbs.       0       1   
Damon Jackson     3736  Aug 08, 1988  5' 11"   71"  145 lbs.       1       4   
Jeremy Jackson    3737  Sep 19, 1982  5' 10"    --  170 lbs.       0       5   
Quinton Jackson   3738  Jun 20, 1978   6' 1"   73"  205 lbs.       0      12   
Eugene Jackson    3739  Sep 23, 1966   5' 8"    --  185 lbs.       1       9   

                 n_win  sig_str_abs_pM  sig_str_def_pct  sig_str_land_pM  \
name                                                                       
Hunter Azure         9            2.08             0.58             3.92   
Virgil Zwicker      15            4.87             0.39             3.34   
Gilbert Yvel        39            1.78             0.50             1.05   
Xie Bin              8            2.60             0.63             2.20   
Wu Yanan            11            4.82             0.51             4.51   
...                ...             ...              ...              ...   
Mike Jackson         0            1.33             0.75             4.25   
Damon Jackson       19            3.25             0.48             2.54   
Jeremy Jackson       9            2.03             0.25             0.71   
Quinton Jackson     37            2.51             0.55             3.25   
Eugene Jackson      15            3.44             0.32             2.19   

                 sig_str_land_pct    stance  sub_avg  td_avg  td_def_pct  \
name                                                                       
Hunter Azure                 0.53  Orthodox      1.3    1.97        0.67   
Virgil Zwicker               0.48       NaN      0.0    1.31        0.50   
Gilbert Yvel                 0.47  Orthodox      1.0    0.00        0.25   
Xie Bin                      0.38  Orthodox      0.0    4.00        0.00   
Wu Yanan                     0.45  Orthodox      0.3    0.61        0.66   
...                           ...       ...      ...     ...         ...   
Mike Jackson                 0.58  Orthodox      0.0    0.95        0.88   
Damon Jackson                0.38    Switch      2.5    2.48        0.35   
Jeremy Jackson               0.27  Orthodox      0.0    1.52        0.58   
Quinton Jackson              0.54  Orthodox      0.1    1.52        0.73   
Eugene Jackson               0.33  Orthodox      2.3    1.91        0.16   

                 td_land_pct Month  Day  Year  
name                                           
Hunter Azure            0.34   Mar  02,  1992  
Virgil Zwicker          0.30   Jun  26,  1982  
Gilbert Yvel            0.00   Jun  30,  1976  
Xie Bin                 0.44   Feb  28,  1998  
Wu Yanan                0.22   Apr  18,  1996  
...                      ...   ...  ...   ...  
Mike Jackson            1.00   Mar  22,  1985  
Damon Jackson           0.33   Aug  08,  1988  
Jeremy Jackson          0.40   Sep  19,  1982  
Quinton Jackson         0.56   Jun  20,  1978  
Eugene Jackson          0.71   Sep  23,  1966  

[3098 rows x 20 columns]

In [15]:
df_fighters_cl["Day"] = df_fighters_cl["Day"].str.replace("[,]"," ")

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  """Entry point for launching an IPython kernel.


In [16]:
# Second, let's convert the month from a word to a number
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Jan"] = 1
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Feb"] = 2
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Mar"] = 3
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Apr"] = 4
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "May"] = 5
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Jun"] = 6
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Jul"] = 7
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Aug"] = 8
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Sep"] = 9
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Oct"] = 10
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Nov"] = 11
df_fighters_cl['Month'].loc[df_fighters_cl['Month'] == "Dec"] = 12
df_fighters_cl

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


index           dob  height reach    weight  n_draw  n_loss  \
name                                                                           
Hunter Azure         0  Mar 02, 1992   5' 8"   69"  145 lbs.       0       2   
Virgil Zwicker       1  Jun 26, 1982   6' 2"   74"  205 lbs.       1       6   
Gilbert Yvel         2  Jun 30, 1976   6' 2"   77"  225 lbs.       1      16   
Xie Bin              3  Feb 28, 1998  5' 10"   72"  145 lbs.       0       3   
Wu Yanan             4  Apr 18, 1996   5' 8"   66"  135 lbs.       0       4   
...                ...           ...     ...   ...       ...     ...     ...   
Mike Jackson      3734  Mar 22, 1985   6' 2"    --  170 lbs.       0       1   
Damon Jackson     3736  Aug 08, 1988  5' 11"   71"  145 lbs.       1       4   
Jeremy Jackson    3737  Sep 19, 1982  5' 10"    --  170 lbs.       0       5   
Quinton Jackson   3738  Jun 20, 1978   6' 1"   73"  205 lbs.       0      12   
Eugene Jackson    3739  Sep 23, 1966   5' 8"    --  185 lbs.       1       9   

                 n_win  sig_str_abs_pM  sig_str_def_pct  sig_str_land_pM  \
name                                                                       
Hunter Azure         9            2.08             0.58             3.92   
Virgil Zwicker      15            4.87             0.39             3.34   
Gilbert Yvel        39            1.78             0.50             1.05   
Xie Bin              8            2.60             0.63             2.20   
Wu Yanan            11            4.82             0.51             4.51   
...                ...             ...              ...              ...   
Mike Jackson         0            1.33             0.75             4.25   
Damon Jackson       19            3.25             0.48             2.54   
Jeremy Jackson       9            2.03             0.25             0.71   
Quinton Jackson     37            2.51             0.55             3.25   
Eugene Jackson      15            3.44             0.32             2.19   

                 sig_str_land_pct    stance  sub_avg  td_avg  td_def_pct  \
name                                                                       
Hunter Azure                 0.53  Orthodox      1.3    1.97        0.67   
Virgil Zwicker               0.48       NaN      0.0    1.31        0.50   
Gilbert Yvel                 0.47  Orthodox      1.0    0.00        0.25   
Xie Bin                      0.38  Orthodox      0.0    4.00        0.00   
Wu Yanan                     0.45  Orthodox      0.3    0.61        0.66   
...                           ...       ...      ...     ...         ...   
Mike Jackson                 0.58  Orthodox      0.0    0.95        0.88   
Damon Jackson                0.38    Switch      2.5    2.48        0.35   
Jeremy Jackson               0.27  Orthodox      0.0    1.52        0.58   
Quinton Jackson              0.54  Orthodox      0.1    1.52        0.73   
Eugene Jackson               0.33  Orthodox      2.3    1.91        0.16   

                 td_land_pct Month  Day  Year  
name                                           
Hunter Azure            0.34     3  02   1992  
Virgil Zwicker          0.30     6  26   1982  
Gilbert Yvel            0.00     6  30   1976  
Xie Bin                 0.44     2  28   1998  
Wu Yanan                0.22     4  18   1996  
...                      ...   ...  ...   ...  
Mike Jackson            1.00     3  22   1985  
Damon Jackson           0.33     8  08   1988  
Jeremy Jackson          0.40     9  19   1982  
Quinton Jackson         0.56     6  20   1978  
Eugene Jackson          0.71     9  23   1966  

[3098 rows x 20 columns]

In [17]:
# Third, let's check if there is any non integer/float values
df_dob_missing = df_fighters_cl[(df_fighters_cl['Month'] 
== "--") | (df_fighters_cl['Day'] == "None") 
 | (df_fighters_cl['Year'] == "None")]
print(len(df_fighters))
print(len(df_st_gr))
print(len(df_fighters)-len(df_st_gr))
print(len(df_dob_missing))
print(len(df_fighters_cl)-len(df_dob_missing))
df_dob_missing.head(5)

3740
642
3098
347
2751


index dob height reach    weight  n_draw  n_loss  n_win  \
name                                                                      
Jarrod Kwitty      17  --  5' 8"    --  185 lbs.       0       2      2   
Lina Kvokov        19  --  5' 9"    --  135 lbs.       0       3      0   
Kyle Kurtz         21  --     --    --  170 lbs.       0       5      7   
Keigo Kunihara     28  --  6' 0"    --  235 lbs.       0       2      3   
Maiju Kujala       32  --     --    --  135 lbs.       0       3      6   

                sig_str_abs_pM  sig_str_def_pct  sig_str_land_pM  \
name                                                               
Jarrod Kwitty             4.40             0.32             0.53   
Lina Kvokov              12.12             0.19             2.31   
Kyle Kurtz                2.95             0.52             1.31   
Keigo Kunihara            1.99             0.57             0.17   
Maiju Kujala              5.50             0.36             2.50   

                sig_str_land_pct    stance  sub_avg  td_avg  td_def_pct  \
name                                                                      
Jarrod Kwitty               0.50       NaN      2.0    8.00         0.0   
Lina Kvokov                 0.30  Orthodox      0.0    0.00         0.0   
Kyle Kurtz                  0.36       NaN      2.5    1.23         0.0   
Keigo Kunihara              0.10  Orthodox      2.5    4.97         1.0   
Maiju Kujala                0.40       NaN      0.0    0.00         0.5   

                td_land_pct Month   Day  Year  
name                                           
Jarrod Kwitty           0.5    --  None  None  
Lina Kvokov             0.0    --  None  None  
Kyle Kurtz              0.2    --  None  None  
Keigo Kunihara          0.4    --  None  None  
Maiju Kujala            0.0    --  None  None

### It seems there is about 347 fighters with missing dob's. We can either fill them out manually from other sources such as ESPN MMA, sherdog, and tapalogy. Or just drop them, and that's I will do

In [18]:
# Create a copy, to prevent error.
df_fighters_cl2 = df_fighters_cl.loc[~((df_fighters_cl['Month'] == "--") | (df_fighters_cl['Day'] == "None") | (df_fighters_cl['Year'] == "None"))].copy()


In [19]:
df_fighters_cl2

index           dob  height reach    weight  n_draw  n_loss  \
name                                                                           
Hunter Azure         0  Mar 02, 1992   5' 8"   69"  145 lbs.       0       2   
Virgil Zwicker       1  Jun 26, 1982   6' 2"   74"  205 lbs.       1       6   
Gilbert Yvel         2  Jun 30, 1976   6' 2"   77"  225 lbs.       1      16   
Xie Bin              3  Feb 28, 1998  5' 10"   72"  145 lbs.       0       3   
Wu Yanan             4  Apr 18, 1996   5' 8"   66"  135 lbs.       0       4   
...                ...           ...     ...   ...       ...     ...     ...   
Mike Jackson      3734  Mar 22, 1985   6' 2"    --  170 lbs.       0       1   
Damon Jackson     3736  Aug 08, 1988  5' 11"   71"  145 lbs.       1       4   
Jeremy Jackson    3737  Sep 19, 1982  5' 10"    --  170 lbs.       0       5   
Quinton Jackson   3738  Jun 20, 1978   6' 1"   73"  205 lbs.       0      12   
Eugene Jackson    3739  Sep 23, 1966   5' 8"    --  185 lbs.       1       9   

                 n_win  sig_str_abs_pM  sig_str_def_pct  sig_str_land_pM  \
name                                                                       
Hunter Azure         9            2.08             0.58             3.92   
Virgil Zwicker      15            4.87             0.39             3.34   
Gilbert Yvel        39            1.78             0.50             1.05   
Xie Bin              8            2.60             0.63             2.20   
Wu Yanan            11            4.82             0.51             4.51   
...                ...             ...              ...              ...   
Mike Jackson         0            1.33             0.75             4.25   
Damon Jackson       19            3.25             0.48             2.54   
Jeremy Jackson       9            2.03             0.25             0.71   
Quinton Jackson     37            2.51             0.55             3.25   
Eugene Jackson      15            3.44             0.32             2.19   

                 sig_str_land_pct    stance  sub_avg  td_avg  td_def_pct  \
name                                                                       
Hunter Azure                 0.53  Orthodox      1.3    1.97        0.67   
Virgil Zwicker               0.48       NaN      0.0    1.31        0.50   
Gilbert Yvel                 0.47  Orthodox      1.0    0.00        0.25   
Xie Bin                      0.38  Orthodox      0.0    4.00        0.00   
Wu Yanan                     0.45  Orthodox      0.3    0.61        0.66   
...                           ...       ...      ...     ...         ...   
Mike Jackson                 0.58  Orthodox      0.0    0.95        0.88   
Damon Jackson                0.38    Switch      2.5    2.48        0.35   
Jeremy Jackson               0.27  Orthodox      0.0    1.52        0.58   
Quinton Jackson              0.54  Orthodox      0.1    1.52        0.73   
Eugene Jackson               0.33  Orthodox      2.3    1.91        0.16   

                 td_land_pct Month  Day  Year  
name                                           
Hunter Azure            0.34     3  02   1992  
Virgil Zwicker          0.30     6  26   1982  
Gilbert Yvel            0.00     6  30   1976  
Xie Bin                 0.44     2  28   1998  
Wu Yanan                0.22     4  18   1996  
...                      ...   ...  ...   ...  
Mike Jackson            1.00     3  22   1985  
Damon Jackson           0.33     8  08   1988  
Jeremy Jackson          0.40     9  19   1982  
Quinton Jackson         0.56     6  20   1978  
Eugene Jackson          0.71     9  23   1966  

[2751 rows x 20 columns]

In [20]:
# Create another column for the DOB, with a specific format.
df_fighters_cl2["DOB"] = df_fighters_cl2["Year"].astype(str)+'/' + df_fighters_cl2["Month"].astype(str)+'/' + df_fighters_cl2["Day"].astype(str)

### I know this is probably the least efficient way to do it, but just as Dr. Rick said "it doesn't need to be pretty, it just needs to work"

In [21]:
# This function converts given date to age
def age(born):
	born = datetime.strptime(born, "%Y/%m/%d ").date()
	today = date.today()
	return today.year - born.year - ((today.month,
									today.day) < (born.month,
													born.day))

df_fighters_cl2['Age'] = df_fighters_cl2['DOB'].apply(age)

In [22]:
# Let's the drop the extra columns from the mess we created
df_fighters_cl2.drop(['dob','Month', 'Day', 'Year','DOB'], axis = 1, inplace=True)

In [23]:
# Finally, we want to see if there is any fighters missing thier height measurement
print(len(df_fighters_cl2.loc[df_fighters_cl2['height'] == '--']))
df_fighters_cl2.loc[df_fighters_cl2['height'] == '--']

11


index height reach    weight  n_draw  n_loss  n_win  \
name                                                                    
Logan Nail          796     --    --  185 lbs.       0       1      1   
Matt Ricehouse     1490     --    --  155 lbs.       0       1      6   
Amador Ramirez     1601     --    --  135 lbs.       0       4      5   
Bryan Travers      1700     --    --  155 lbs.       0       4     15   
Maka Watson        1983     --    --  155 lbs.       0       2      4   
Hiroshi Izumi      2290     --    --  205 lbs.       0       2      4   
TJ Cook            2643     --    --  205 lbs.       0       5     13   
Joe Duarte         2839     --    --  155 lbs.       0       4     10   
Neal Ewing         2993     --    --  185 lbs.       0       0      6   
Edward Faaloloto   3191     --   70"  155 lbs.       0       5      2   
Lee Higgins        3482     --    --  155 lbs.       0       1      2   

                  sig_str_abs_pM  sig_str_def_pct  sig_str_land_pM  \
name                                                                 
Logan Nail                  2.27             0.39             1.93   
Matt Ricehouse              4.80             0.47             3.70   
Amador Ramirez              2.07             0.69             4.93   
Bryan Travers               3.93             0.55             2.33   
Maka Watson                 1.60             0.22             0.93   
Hiroshi Izumi               2.65             0.66             1.95   
TJ Cook                     3.18             0.54             2.30   
Joe Duarte                  4.00             0.53             2.27   
Neal Ewing                  1.93             0.48             2.27   
Edward Faaloloto            6.25             0.44             2.28   
Lee Higgins                 3.68             0.40             1.02   

                  sig_str_land_pct    stance  sub_avg  td_avg  td_def_pct  \
name                                                                        
Logan Nail                    0.51       NaN      0.0    0.00        0.37   
Matt Ricehouse                0.44       NaN      0.0    1.00        0.81   
Amador Ramirez                0.51       NaN      0.0    1.00        0.00   
Bryan Travers                 0.48       NaN      0.0    2.00        0.63   
Maka Watson                   0.37       NaN      0.0    2.00        0.33   
Hiroshi Izumi                 0.37  Orthodox      0.5    3.35        1.00   
TJ Cook                       0.47       NaN      0.0    1.01        0.00   
Joe Duarte                    0.38       NaN      1.0    3.00        0.69   
Neal Ewing                    0.60       NaN      0.0    5.00        0.00   
Edward Faaloloto              0.32       NaN      0.0    1.01        0.33   
Lee Higgins                   0.26       NaN      0.0    0.00        0.00   

                  td_land_pct  Age  
name                                
Logan Nail               0.00   32  
Matt Ricehouse           0.22   34  
Amador Ramirez           0.33   31  
Bryan Travers            0.28   38  
Maka Watson              1.00   37  
Hiroshi Izumi            0.70   39  
TJ Cook                  0.50   39  
Joe Duarte               0.50   45  
Neal Ewing               0.62   36  
Edward Faaloloto         0.25   37  
Lee Higgins              0.00   41

### It seems we have 11 fighters that are missing thier heights, and since it is only a slight number. I think we can manually input it, fairly easy. 
### But first we need seperate the height, because it will make handling inputting a "qutation mark" much easier  

In [24]:
# Just like with the "DOB" column, I need to seperate the height
df_fighters_cl2[["feet", "inch"]] = df_fighters_cl2.height.str.split(" ", expand=True)

In [25]:
# Remove the ' from feet, and the " fro inch columns
df_fighters_cl2["feet"] = df_fighters_cl2["feet"].str.replace("'","", regex=True)
df_fighters_cl2["inch"] = df_fighters_cl2["inch"].str.replace('"',"", regex=True)

In [26]:
# Input the heights manually from shredog, and tapalogy.
df_fighters_cl2.loc['Logan Nail', "feet"] = '6'
df_fighters_cl2.loc['Logan Nail', "inch"] = '3'
df_fighters_cl2.loc['Matt Ricehouse', "feet"] = '5'
df_fighters_cl2.loc['Matt Ricehouse', "inch"] = '10'
df_fighters_cl2.loc['Amador Ramirez', "feet"] = '5'
df_fighters_cl2.loc['Amador Ramirez', "inch"] = '7'
df_fighters_cl2.loc['Bryan Travers', "feet"] = '5'
df_fighters_cl2.loc['Bryan Travers', "inch"] = '10'
df_fighters_cl2.loc['Maka Watson', "feet"] = '5'
df_fighters_cl2.loc['Maka Watson', "inch"] = '9'
df_fighters_cl2.loc['Hiroshi Izumi', "feet"] = '5'
df_fighters_cl2.loc['Hiroshi Izumi', "inch"] = '8'
df_fighters_cl2.loc['TJ Cook', "feet"] = '5'
df_fighters_cl2.loc['TJ Cook', "inch"] = '10'
df_fighters_cl2.loc['Joe Duarte', "feet"] = '5'
df_fighters_cl2.loc['Joe Duarte', "inch"] = '9'
df_fighters_cl2.loc['Neal Ewing', "feet"] = '6'
df_fighters_cl2.loc['Neal Ewing', "inch"] = '0'
df_fighters_cl2.loc['Edward Faaloloto', "feet"] = '5'
df_fighters_cl2.loc['Edward Faaloloto', "inch"] = '9'
df_fighters_cl2.loc['Lee Higgins', "feet"] = '5'
df_fighters_cl2.loc['Lee Higgins', "inch"] = '10'

In [27]:
# Convert the column types to "integer", to be able to apply mathematical equations
df_fighters_cl2["feet"]= df_fighters_cl2["feet"].astype(int)
df_fighters_cl2["inch"]= df_fighters_cl2["inch"].astype(int)

### Now, I would like to convert the height to metric units, because I think it's better to handle and it has no impact on the overall image

In [28]:
# Create a new column with the height in cm, and apply conversion 
df_fighters_cl2["height_cm"] = round(df_fighters_cl2["feet"]*30.48 + df_fighters_cl2["inch"]*2.54)

### We are getting closer to have "clean" data, after dealing with the height we deal with the reach

In [29]:
# Now we want to clean the "reach" column
print(len(df_fighters_cl2.loc[df_fighters_cl2['reach'] == '--']))
df_fighters_cl2.loc[df_fighters_cl2['reach'] == '--']

925


index  height reach    weight  n_draw  n_loss  n_win  \
name                                                                      
Igor Vovchanchyn       5   5' 8"    --  205 lbs.       0      10     55   
Austin Tweedy          8   5' 9"    --  155 lbs.       0       2      8   
Eldari Kurtanidze     22   5' 6"    --  235 lbs.       0       1      0   
Tsuyoshi Kurihara     23  5' 10"    --  185 lbs.       2       8     10   
Leo Kuntz             26  5' 10"    --  155 lbs.       1       4     18   
...                  ...     ...   ...       ...     ...     ...    ...   
Nate James          3728   6' 0"    --  185 lbs.       1       8     14   
Justin James        3729  5' 10"    --  155 lbs.       0       7     13   
Mike Jackson        3734   6' 2"    --  170 lbs.       0       1      0   
Jeremy Jackson      3737  5' 10"    --  170 lbs.       0       5      9   
Eugene Jackson      3739   5' 8"    --  185 lbs.       1       9     15   

                   sig_str_abs_pM  sig_str_def_pct  sig_str_land_pM  \
name                                                                  
Igor Vovchanchyn             0.97             0.63             1.59   
Austin Tweedy                6.43             0.00             4.29   
Eldari Kurtanidze            3.72             0.60             0.47   
Tsuyoshi Kurihara           12.00             0.50             0.00   
Leo Kuntz                    3.80             0.45             1.68   
...                           ...              ...              ...   
Nate James                   2.67             0.28             1.33   
Justin James                 2.68             0.50             1.34   
Mike Jackson                 1.33             0.75             4.25   
Jeremy Jackson               2.03             0.25             0.71   
Eugene Jackson               3.44             0.32             2.19   

                   sig_str_land_pct    stance  sub_avg  td_avg  td_def_pct  \
name                                                                         
Igor Vovchanchyn               0.49  Orthodox      0.5    1.19        0.50   
Austin Tweedy                  0.66  Orthodox      0.0    0.00        0.00   
Eldari Kurtanidze              0.20  Orthodox      0.0    0.00        0.00   
Tsuyoshi Kurihara              0.00  Orthodox      0.0    0.00        0.00   
Leo Kuntz                      0.32  Orthodox      0.7    1.99        0.33   
...                             ...       ...      ...     ...         ...   
Nate James                     0.42       NaN      0.0    0.00        0.44   
Justin James                   0.36  Orthodox      1.6    0.00        0.50   
Mike Jackson                   0.58  Orthodox      0.0    0.95        0.88   
Jeremy Jackson                 0.27  Orthodox      0.0    1.52        0.58   
Eugene Jackson                 0.33  Orthodox      2.3    1.91        0.16   

                   td_land_pct  Age  feet  inch  height_cm  
name                                                        
Igor Vovchanchyn          0.66   48     5     8      173.0  
Austin Tweedy             0.00   31     5     9      175.0  
Eldari Kurtanidze         0.00   49     5     6      168.0  
Tsuyoshi Kurihara         0.00   48     5    10      178.0  
Leo Kuntz                 0.25   38     5    10      178.0  
...                        ...  ...   ...   ...        ...  
Nate James                0.00   43     6     0      183.0  
Justin James              0.00   45     5    10      178.0  
Mike Jackson              1.00   36     6     2      188.0  
Jeremy Jackson            0.40   39     5    10      178.0  
Eugene Jackson            0.71   55     5     8      173.0  

[925 rows x 20 columns]

### Now there is multiple ways to handle the missing "reach" values:
    - Delete the rows
    - Replace with the mean
    - Replace with the height
### The average human has the same reach as their height, that ratio is called the "ape index". So, in this case a good solution would to assume all the fighters missing thier reach have an "ape index" of 1. Or in other words their reach is equal to thier height


In [30]:
# Input the reaches manually from shredog, and tapalogy.
df_fighters_cl2.loc['Logan Nail', "reach"] = '80"'

df_fighters_cl2.loc['Matt Ricehouse', "reach"] = '72"'

df_fighters_cl2.loc['Amador Ramirez', "reach"] = '68"'

df_fighters_cl2.loc['Bryan Travers', "reach"] = '69"'

df_fighters_cl2.loc['Maka Watson', "reach"] = '69"'

df_fighters_cl2.loc['Hiroshi Izumi', "reach"] = '68"'

df_fighters_cl2.loc['TJ Cook', "reach"] = '72"'

df_fighters_cl2.loc['Joe Duarte', "reach"] = '68"'

df_fighters_cl2.loc['Neal Ewing', "reach"] = '76"'

df_fighters_cl2.loc['Lee Higgins', "reach"] = '74"'


In [31]:
# Now we can handle the column as it is, in a very similar way as the in the heights
df_fighters_cl2["reach"] = df_fighters_cl2["reach"].str.replace('"',"", regex=True)

In [32]:
# Create a column for the height in inches, to replace with the reach
df_fighters_cl2["height_inch"] = round(df_fighters_cl2["feet"]*12 + df_fighters_cl2["inch"]*1)

In [33]:
# Now, we need to replace the reach values with the height in inches.
df_fighters_cl2['reach'] = np.where(df_fighters_cl2['reach'] == "--", df_fighters_cl2['height_inch'], df_fighters_cl2['reach'])

In [34]:
# Convert the column types to "ing", to be able to apply math equations
df_fighters_cl2["reach"]= df_fighters_cl2["reach"].astype(int)

In [35]:
# Again, just like in the heights. We convert the reach to cm
df_fighters_cl2["reach_cm"] = round(df_fighters_cl2["reach"]*2.54)

### We can create that "ape ratio" we talked about, and althought it's redudant and we will probably get rid of in the feature engineering process. But, it would be nice to visuzalize

In [36]:
# Create new column for ape index
df_fighters_cl2["ape_index"] = (df_fighters_cl2["reach_cm"]/df_fighters_cl2["height_cm"])

### Finally, we can now get to handeling the stance column which is the last column in our to do list

In [37]:
# Lets see how many figters we have with no values in the stance column
df_fighters_cl2['stance'].isnull().sum()

204

### In this one the line for what to do is very vage, for now we will just drop that column.

In [38]:
df_fighters_cl2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2751 entries, Hunter Azure to Eugene Jackson
Data columns (total 23 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   index             2751 non-null   int64  
 1   height            2751 non-null   object 
 2   reach             2751 non-null   int64  
 3   weight            2751 non-null   object 
 4   n_draw            2751 non-null   int64  
 5   n_loss            2751 non-null   int64  
 6   n_win             2751 non-null   int64  
 7   sig_str_abs_pM    2751 non-null   float64
 8   sig_str_def_pct   2751 non-null   float64
 9   sig_str_land_pM   2751 non-null   float64
 10  sig_str_land_pct  2751 non-null   float64
 11  stance            2547 non-null   object 
 12  sub_avg           2751 non-null   float64
 13  td_avg            2751 non-null   float64
 14  td_def_pct        2751 non-null   float64
 15  td_land_pct       2751 non-null   float64
 16  Age               2751 non

In [39]:
# Let's add a new column to quantify a winning ratio
df_fighters_cl2["win_percentage"] = ((df_fighters_cl2["n_win"])/(df_fighters_cl2["n_win"] + df_fighters_cl2["n_loss"] + df_fighters_cl2["n_draw"]))

In [40]:
# Let's add a new column to quanitfy the experience (number of fights)
df_fighters_cl2["Number_of_fights"] = ((df_fighters_cl2["n_win"]+df_fighters_cl2["n_loss"]+df_fighters_cl2["n_draw"]))

### Another problem came during the data viz, which is the weight and we need to solve it here.
### We need to convert the weight to countinous data and then convert back to categorical

In [41]:
# Remove the lbs. to prepare to convert to integer
df_fighters_cl2["weight"] = df_fighters_cl2["weight"].str.replace('lbs.',"", regex=True)

In [42]:
# Convert the weight to integer
df_fighters_cl2["weight"]= df_fighters_cl2["weight"].astype(int)

In [43]:
# Function to recreat the weight classes
def weight_cat(x):
    if x <= 115:
        return "Strawweight 115 lb"
    elif 116 <= x <= 125:
        return "Flyweight 125 lb"
    elif 126 <= x <= 135:
        return "Bantamweight 135 lb"
    elif 136 <= x <= 145:
        return "Featherweight 145 lb"
    elif 146 <= x <= 155:
        return "Lightweight 155 lb"
    elif 156 <= x <= 170:
        return "Welterweight 170 lb"
    elif 171 <= x <= 185:
        return "Middleweight 185 lb"
    elif 186 <= x <= 205:
        return "Light Heavyweight 205 lb"
    elif 206 <= x <= 265:
        return "Heavyweight 265 lb"
    else:
        return "Super Heavyweight"

In [44]:
# Applying the function to the data
df_fighters_cl2["weight_class"] = df_fighters_cl2['weight'].apply(weight_cat)

In [45]:
# Finally, let's get rid of all the unnecessary columns
df_fighters_cl2.drop(['height','feet', 'inch', 'height_inch','stance'], axis = 1, inplace=True)

In [46]:
# Create another index
df_fighters_cl2['Number'] = range(1, len(df_fighters_cl2) + 1)

In [47]:
# Drop columns before finishing the last details
df_fighters_cl2.drop(['index','n_draw', 'n_loss', 'n_win','reach'], axis = 1, inplace=True)

In [48]:
df = df_fighters_cl2

### Final process before saving the file, is to rename and rearrange the columns

In [49]:
df.head()

weight  sig_str_abs_pM  sig_str_def_pct  sig_str_land_pM  \
name                                                                       
Hunter Azure       145            2.08             0.58             3.92   
Virgil Zwicker     205            4.87             0.39             3.34   
Gilbert Yvel       225            1.78             0.50             1.05   
Xie Bin            145            2.60             0.63             2.20   
Wu Yanan           135            4.82             0.51             4.51   

                sig_str_land_pct  sub_avg  td_avg  td_def_pct  td_land_pct  \
name                                                                         
Hunter Azure                0.53      1.3    1.97        0.67         0.34   
Virgil Zwicker              0.48      0.0    1.31        0.50         0.30   
Gilbert Yvel                0.47      1.0    0.00        0.25         0.00   
Xie Bin                     0.38      0.0    4.00        0.00         0.44   
Wu Yanan                    0.45      0.3    0.61        0.66         0.22   

                Age  height_cm  reach_cm  ape_index  win_percentage  \
name                                                                  
Hunter Azure     30      173.0     175.0   1.011561        0.818182   
Virgil Zwicker   39      188.0     188.0   1.000000        0.681818   
Gilbert Yvel     45      188.0     196.0   1.042553        0.696429   
Xie Bin          24      178.0     183.0   1.028090        0.727273   
Wu Yanan         25      173.0     168.0   0.971098        0.733333   

                Number_of_fights              weight_class  Number  
name                                                                
Hunter Azure                  11      Featherweight 145 lb       1  
Virgil Zwicker                22  Light Heavyweight 205 lb       2  
Gilbert Yvel                  56        Heavyweight 265 lb       3  
Xie Bin                       11      Featherweight 145 lb       4  
Wu Yanan                      15       Bantamweight 135 lb       5

In [50]:
# Renaming columns
df.rename(columns={'sig_str_abs_pM': "D_SApM", 
'sig_str_def_pct':"D_StrPct", 
'sig_str_land_pM':"O_SLpM",
'sig_str_land_pct':"O_StrPct",
'sub_avg':"O_SubAvg",
'td_avg':"O_TdAvg",
'td_def_pct':"D_TdPct",
'td_land_pct':"O_TdAcc",
'Number':'index'}, inplace=True)

In [51]:
# Rearranging columns
df = df[['index', 'Age','weight_class','height_cm', 'reach_cm', 
       'ape_index','Number_of_fights','win_percentage', 'O_SLpM', 'O_StrPct', 
       'D_SApM', 'D_StrPct', 'O_TdAvg',  'O_TdAcc', 'O_SubAvg', 'D_TdPct',
       'weight']]

In [52]:
df.head()

index  Age              weight_class  height_cm  reach_cm  \
name                                                                        
Hunter Azure        1   30      Featherweight 145 lb      173.0     175.0   
Virgil Zwicker      2   39  Light Heavyweight 205 lb      188.0     188.0   
Gilbert Yvel        3   45        Heavyweight 265 lb      188.0     196.0   
Xie Bin             4   24      Featherweight 145 lb      178.0     183.0   
Wu Yanan            5   25       Bantamweight 135 lb      173.0     168.0   

                ape_index  Number_of_fights  win_percentage  O_SLpM  O_StrPct  \
name                                                                            
Hunter Azure     1.011561                11        0.818182    3.92      0.53   
Virgil Zwicker   1.000000                22        0.681818    3.34      0.48   
Gilbert Yvel     1.042553                56        0.696429    1.05      0.47   
Xie Bin          1.028090                11        0.727273    2.20      0.38   
Wu Yanan         0.971098                15        0.733333    4.51      0.45   

                D_SApM  D_StrPct  O_TdAvg  O_TdAcc  O_SubAvg  D_TdPct  weight  
name                                                                           
Hunter Azure      2.08      0.58     1.97     0.34       1.3     0.67     145  
Virgil Zwicker    4.87      0.39     1.31     0.30       0.0     0.50     205  
Gilbert Yvel      1.78      0.50     0.00     0.00       1.0     0.25     225  
Xie Bin           2.60      0.63     4.00     0.44       0.0     0.00     145  
Wu Yanan          4.82      0.51     0.61     0.22       0.3     0.66     135

In [53]:
# Save clean file to csv
df.to_csv("/Users/ksah4444/Documents/UFC_ML_Prediction/data/Clean_Data/df_fighters.csv")